mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java

```diff
@@ -19,19 +19,133 @@
 
 package org.apache.james.mailbox.events.delivery;
 
-import java.util.Collection;
+import static org.apache.james.mailbox.events.delivery.EventDelivery.PermanentFailureHandler.NO_HANDLER;
+import static org.apache.james.mailbox.events.delivery.EventDelivery.Retryer.NO_RETRYER;
+
+import java.time.Duration;
 
 import org.apache.james.mailbox.Event;
 import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.events.EventDeadLetters;
+import org.apache.james.mailbox.events.Group;
+import org.apache.james.mailbox.events.RetryBackoffConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
 public interface EventDelivery {
+
+    class DeliveryOption {
+        public static DeliveryOption of(Retryer retrier, PermanentFailureHandler permanentFailureHandler) {
+            return new DeliveryOption(retrier, permanentFailureHandler);
+        }
+
+        public static DeliveryOption none() {
+            return new DeliveryOption(NO_RETRYER, NO_HANDLER);
+        }
+
+        private final Retryer retrier;
+        private final PermanentFailureHandler permanentFailureHandler;
+
+        private DeliveryOption(Retryer retrier, PermanentFailureHandler permanentFailureHandler) {
+            this.retrier = retrier;
+            this.permanentFailureHandler = permanentFailureHandler;
+        }
+
+        Retryer getRetrier() {
+            return retrier;
+        }
+
+        PermanentFailureHandler getPermanentFailureHandler() {
+            return permanentFailureHandler;
+        }
+    }
+
+
+    interface Retryer {
+
+        Retryer NO_RETRYER = (executionResult, event) -> executionResult;
+
+        class BackoffRetryer implements EventDelivery.Retryer {
+
+            public static BackoffRetryer of(RetryBackoffConfiguration retryBackoff, MailboxListener mailboxListener) {
+                return new BackoffRetryer(retryBackoff, mailboxListener);
+            }
+
+            private static final Logger LOGGER = LoggerFactory.getLogger(BackoffRetryer.class);
+            private static final Duration MAX_BACKOFF = Duration.ofMillis(Long.MAX_VALUE);
```

Comment:
`Long.MAX_VALUE` seems big, is their reason why we want such a delay?

Reply:
As I understand it, it is a higher threshold for wait `delay*jitter^retry`

We do not care about such a delay, that would anyway be lower than this and do not specify this parameter in reactor `retryBackOff` method signature. Passing an inifinite duration ensures that you do not mess up with other parameters.

This can safely be done BTW